File: internal/rediscache/redis_cache.go
package rediscache

import (
	"encoding/json"
	"github.com/hpifu/go-account/internal/c"
	"time"
	"github.com/go-redis/redis"
	"github.com/hpifu/go-account/internal/mysqldb"
)

type Option struct {
	Address            string
	Timeout            time.Duration
	Retries            int
	PoolSize           int
	Password           string
	DB                 int
	TokenExpiration    time.Duration
	AuthCodeExpiration time.Duration
}

type RedisCache struct {
	client *redis.Client
	option *Option
}

type Account struct {
	ID        int      `json:"id,omitempty"`
	Email     string   `json:"email,omitempty"`
	Phone     string   `json:"phone,omitempty"`
	FirstName string   `json:"firstName,omitempty"`
	LastName  string   `json:"lastName,omitempty"`
	Birthday  string   `json:"birthday,omitempty"`
	Password  string   `json:"password,omitempty"`
	Gender    c.Gender `json:"gender"`
}

func NewAccount(account *mysqldb.Account) *Account {
	return &Account{
		ID:        account.ID,
		Email:     account.Email,
		Phone:     account.Phone,
		FirstName: account.FirstName,
		LastName:  account.LastName,
		Password:  account.Password,
		Birthday:  account.Birthday.Format("2006-01-02"),
		Gender:    account.Gender,
	}
}

func NewRedisCache(option *Option) (*RedisCache, error) {
	client := redis.NewClient(&redis.Options{
		Addr:         option.Address,
		DialTimeout:  option.Timeout,
		ReadTimeout:  option.Timeout,
		WriteTimeout: option.Timeout,
		MaxRetries:   option.Retries,
		PoolSize:     option.PoolSize,
		Password:     option.Password,
		DB:           option.DB,
	})

	if err := client.Ping().Err(); err != nil {
		return nil, err
	}

	return &RedisCache{
		client: client,
		option: option,
	}, nil
}

func (rc *RedisCache) SetAuthCode(key string, code string) error {
	return rc.client.Set("ac_"+key, code, rc.option.AuthCodeExpiration).Err()
}

func (rc *RedisCache) DelAuthCode(key string) error {
	return rc.client.Del("ac_" + key).Err()
}

func (rc *RedisCache) GetAuthCode(key string) (string, error) {
	buf, err := rc.client.Get("ac_" + key).Result()
	if err == redis.Nil {
		return "", nil
	}
	if err != nil {
		return "", err
	}

	return string(buf), nil
}

func (rc *RedisCache) SetAccount(token string, account *Account) error {
	buf, err := json.Marshal(account)
	if err != nil {
		return err
	}

	return rc.client.Set(token, buf, rc.option.TokenExpiration).Err()
}

func (rc *RedisCache) DelAccount(token string) error {
	return rc.client.Del(token).Err()
}

func (rc *RedisCache) GetAccount(token string) (*Account, error) {
	buf, err := rc.client.Get(token).Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	account := &Account{}
	if err := json.Unmarshal([]byte(buf), account); err != nil {
		return nil, err
	}

	return account, nil
}

File: features/steps/rediscache.py
#!/usr/bin/env python3

from behave import *
from hamcrest import *
import json


@given('rediscache.token 创建 token: "{token:str}", phone: "{phone:str}", email: "{email:str}", firstname: "{firstname:str}", lastname: "{lastname:str}", birthday: "{birthday:str}", gender: {gender:int}')
def step_impl(context, token, phone, email, firstname, lastname, birthday, gender):
    account = {
        "phone": phone,
        "email": email,
        "firstName": firstname,
        "lastName": lastname,
        "birthday": birthday,
        "gender": gender,
    }
    context.redis_client.set(token, json.dumps(account))


@then('检查 rediscache.token, 存在记录 phone: "{phone:str}", email: "{email:str}", firstname: "{firstname:str}", lastname: "{lastname:str}", birthday: "{birthday:str}", gender: {gender:int}')
def step_impl(context, phone, email, firstname, lastname, birthday, gender):
    res = context.redis_client.get(context.token)
    account = json.loads(res)
    assert_that(account["phone"], equal_to(phone))
    assert_that(account["email"], equal_to(email))
    assert_that(account["firstName"], equal_to(firstname))
    assert_that(account["lastName"], equal_to(lastname))
    assert_that(account["birthday"], equal_to(birthday))
    assert_that(account["gender"], equal_to(gender))


@then('检查 rediscache.token, 不存在记录 token: "{token:str}"')
def step_impl(context, token):
    res = context.redis_client.get(token)
    print(res)
    assert_that(res, equal_to(None))


@given('rediscache.authcode 创建 key: "{key:str}", code: "{code:str}"')
def step_impl(context, key, code):
    context.redis_client.set("ac_"+key, code)


@then('检查 rediscache.authcode, 存在记录 key: "{key:str}"')
def step_impl(context, key):
    res = context.redis_client.get("ac_" + key)
    print(res)
    assert_that(len(res) == 6)
